### DISCUSSION TITLE
Winget-AutoUpdate on Windows server?

### DISCUSSION THREAD

**Discussion by dbennett2843** [Romanitho/Winget-AutoUpdate#738] [Q&A]
Can Winget-AutoUpdate work on any version of Windows Server 2019 or 2022?

**Comment by Romanitho** [comment]
Hi,

I would say yes if Winget is installed on the server. But perhaps it’s not recommended, except for RDS servers?

**Reply by dbennett2843** [reply]
That's exactly what I'm thinking...  just didn't know if there was any history about whether it would work or not :-)

**Reply by Romanitho** [reply]
I believe this topic has already been discussed. However, WAU takes care of prerequisites. So, it should work even on servers. 😅

**Reply by dbennett2843** [reply]
It does install on Windows Server 2022, but it won't run as it reports back a metered connection. I checked the ethernet NIC and metered is OFF

Any ideas?

Log reads

#################################################################
#    10/16/2024 - CHECK FOR APP UPDATES (System context - Connected user)
#################################################################
17:59:07 - Notification Level: None. Notification Language: English
17:59:07 - Checking internet connection...
17:59:07 - Connected !
17:59:07 - Metered connection detected.
17:59:07 - WAU is configured to bypass update checking on metered connection
17:59:07 - End of process!

**Reply by Romanitho** [reply]
It is possible to bypass metered connection checks. Furthermore, for servers, it is not an issue; you should never be on mobile data sharing. 😅
`DONOTRUNONMETERED=0`

**Reply by KnifMelti** [reply]
It's a standard issue with our VMware virtual servers

**Reply by MTC-Lab** [reply]
Same Problem here. All Virtual machines (Servers) are affected.
Metered Connection Detection is NOT working in WAU.
All virtual Interfaces (LAN) are detected as metered by WAU.
Is there any Solution? How is it possible to tun this wrong metered Connection detection off?
How, where? I´ve not understood how i can add parameters to the scripts.
Thank you for your hints.

**Reply by KnifMelti** [reply]
When installing with msi
DONOTRUNONMETERED=0 or Edit the regsetting

**Reply by MTC-Lab** [reply]
OK, this worked.  REG Setting:
-- cut --
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SOFTWARE\Romanitho\Winget-AutoUpdate]
"WAU_DoNotRunOnMetered"=dword:00000000
-- cut --

**Comment by MTC-Lab** [comment]
Yes, it is running on Servers also.
I have it successful implemented at round about >30 Servers under VMWare/HyperV and Windows Server 2019 or 2022.

Install "WinGet" on the Servers via powershell Script
Install WAU per MSI
Set WAU_DoNotRunOnMetered"=dword:00000000 in the registry
Edit the Exclusion List in the Config file (at best delete everything listed there)
Execute the Task in the Task Scheduler.
Check the Logfile

**Reply by Romanitho** [reply]
just add DONOTRUNONMETERED=0 to your msi install command in your script:
`msiexec /i WAU.msi /qn DONOTRUNONMETERED=0`

**Reply by sergiu1223** [reply]
i have a problem with the "default_excluded_apps.txt" it reapears on a weekly basis....

**Comment by sergiu1223** [comment]
Hello everyone,
I have a question regarding the WAU installation on Windows Server 2019.

I'm experiencing an issue with the "default_excluded_apps.txt" file in the configuration folder. It seems that some vital updates are being skipped because of it.

I tried deleting the file, but it keeps reappearing. I also tried clearing its content, but that didn't work either.

Does anyone know how to manage or disable this file properly?

Thanks in advance!

**Reply by Romanitho** [reply]
The default is given as is. Create your own `excluded_apps.txt` with no apps if needed.

**Comment by MTC-Lab** [comment]
The Problem is:

yes, all users will edit the excluded_apps.txt - also done by my side!
(deleted everything inside this file)

BUT with each Update of Romanitho/Winget-AutoUpdate,
MY FILE will be overwritten with your defaults!
Result is: At the Clients almost nothing will be updated.

Please make it this way:
If there is an existing edited File, don´t touch it by an update, also if the file is empty.

And of course your default should be "nothing is excluded"! (File should be empty by default).
The topic of your project is to update Software!
If almost all major critical Softwares are exluded by default, this is nonsense.

--


The default is given as is. Create your own excluded_apps.txt with no apps if needed.

**Reply by Romanitho** [reply]
Regarding your comment: Are you referring to the `excluded_apps.txt` file?
- The default exclusions in that file are in place because the applications listed generally have their own dedicated auto-update systems which are often more robust and configurable than what Winget can offer. This is done to avoid potential conflicts or redundant update processes.
- The `default_excluded_apps.txt` file provided is the default configuration. If you wish to have a different behavior (e.g., updating all applications), the recommended approach is to create your own `excluded_apps.txt` file and leave it empty. This custom file will not be overwritten during updates.
- If you feel this approach doesn't align with your needs, you are welcome to explore other projects that might better suit your requirements. However, the current design aims to balance comprehensive updating with avoiding interference with applications that manage their updates effectively.

**Reply by KnifMelti** [reply]
![image](https://github.com/user-attachments/assets/19f1cc7a-b985-4436-8789-4188356c26e3)

**Reply by dbennett2843** [reply]
Quick clarifying question... if I create my own excluded_apps.txt, is the default_excluded_apps.txt ignored, or are both merged into a single list of excluded apps?

**Reply by KnifMelti** [reply]
Ignored

**Reply by Champ895** [reply]
Do we add the custom excluded_apps.txt under config or on the main winget folder under program files? I have it under config and it's still ignoring and using the default one

**Reply by Champ895** [reply]
Looks like it should be under the main folder

**Comment by Champ895** [comment]
Hey guys, how are you deploying this to your servers? GPO? and will this work on AVD (azure virtual desktop)? I have been having a hard time getting this working on system context. It seems to not work or run. If one of you can help me with how you have it deployed, that would be amazing!

I have the auto-updater with the MS store and admx template working find for other endpoints. It is just "not applicable" to AVD through Intune. If you have a script that I can use to get this working, that would be amazing. Thank you!

**Reply by KnifMelti** [reply]
> I have the auto-updater with the MS store and admx template working find for other endpoints. It is just "not applicable" to AVD through Intune.

If this is what you mean, that's another repo than this original (a fork):
https://apps.microsoft.com/detail/xp89bsk82w9j28?hl=en-US&gl=US

**Reply by Champ895** [reply]
Yeah, and that one does not work with AVD. So I am looking for one that will work with AVD and can be deployed through Intune. I want to target our AVD VMs first and then servers next. If you can help with an easy to use script for deployment through Intune with just a blacklist of apps not to touch, that would be amazing.

For Servers, I would need to target using a GPO

**Reply by KnifMelti** [reply]
For starters (using AI):
https://claude.ai/share/38262acf-2fa1-4bcd-9a2b-f02e04cd2a98

Rember in Intune to run the x64 PowerShell:
`%systemroot%\SysNative\WindowsPowershell\v1.0\PowerShell.exe`

This will give you a standard installation with an excluded_apps.txt:
![image](https://github.com/user-attachments/assets/3ba52357-f7fd-4ee0-a7c0-79e471b3b13f)

I recommend you adapt the installation with your own parameters:
https://github.com/Romanitho/Winget-AutoUpdate?tab=readme-ov-file#advanced-installation

**Reply by Champ895** [reply]
Thanks - I will look into it further. The one you wrote using AI, it doesn't have a scheduled run time of daily or when it should actually update all apps. It will be run on system contect from intune with a command of powershell.exe -ExecutionPolicy Bypass -File Deploy-WAU.ps1


How does this look?
#Requires -RunAsAdministrator

param(
    [string[]]$BlacklistedApps = @(
        "Microsoft.Office.Desktop",
        "NitroSoftware.NitroPro"
    )
)

# Config variables
$TempPath = "$env:TEMP\WAU_Deploy"
$GitHubRepo = "Romanitho/Winget-AutoUpdate"
$LogFile = "$env:ProgramData\WAU_Deploy.log"

function Write-Log {
    param([string]$Message)
    $Timestamp = Get-Date -Format "yyyy-MM-dd HH:mm:ss"
    "$Timestamp - $Message" | Tee-Object -FilePath $LogFile -Append
}

try {
    Write-Log "Starting Winget-AutoUpdate deployment"

    # Create temp directory
    if (!(Test-Path $TempPath)) {
        New-Item -ItemType Directory -Path $TempPath -Force | Out-Null
    }

    # Get latest release info from GitHub API
    Write-Log "Fetching latest release information from GitHub"
    $ApiUrl = "https://api.github.com/repos/$GitHubRepo/releases/latest"
    $Release = Invoke-RestMethod -Uri $ApiUrl -UseBasicParsing

    # Find MSI download URL
    $MsiAsset = $Release.assets | Where-Object { $_.name -like "*.msi" }
    if (!$MsiAsset) {
        throw "MSI file not found in latest release"
    }

    $MsiUrl = $MsiAsset.browser_download_url
    $MsiPath = Join-Path $TempPath $MsiAsset.name

    Write-Log "Downloading MSI: $($MsiAsset.name)"
    Invoke-WebRequest -Uri $MsiUrl -OutFile $MsiPath -UseBasicParsing

    # Install WAU silently
    Write-Log "Installing Winget-AutoUpdate silently"
    $InstallArgs = "/i `"$MsiPath`" /quiet /norestart"
    $Process = Start-Process -FilePath "msiexec.exe" -ArgumentList $InstallArgs -Wait -PassThru

    if ($Process.ExitCode -ne 0) {
        throw "Installation failed with exit code: $($Process.ExitCode)"
    }

    # Configure blacklist file
    if ($BlacklistedApps.Count -gt 0) {
        Write-Log "Configuring blacklist"
        $WAUPath = "${env:ProgramFiles}\Winget-AutoUpdate"
        $BlacklistPath = Join-Path $WAUPath "excluded_apps.txt"
        $BlacklistedApps | Out-File -FilePath $BlacklistPath -Encoding UTF8 -Force
        Write-Log "Blacklisted apps: $($BlacklistedApps -join ', ')"
    }

    # Create WAUConfig.xml with settings
    Write-Log "Creating WAUConfig.xml with schedule and notification settings"
    $ConfigPath = Join-Path $WAUPath "WAUConfig.xml"
    $xmlContent = @"
<WAUConfig>
  <Schedule>
    <UpdateFrequency>Daily</UpdateFrequency>
    <UpdateAtTime>12:00</UpdateAtTime>
    <UpdateAtLogon>True</UpdateAtLogon>
    <ReinstallOnPolicyUpdate>True</ReinstallOnPolicyUpdate>
  </Schedule>
  <Notifications>
    <Level>SuccessOnly</Level>
  </Notifications>
  <Blacklist>
    <AppID>Microsoft.Office.Desktop</AppID>
    <AppID>NitroSoftware.NitroPro</AppID>
  </Blacklist>
  <DesktopShortcut>False</DesktopShortcut>
  <StartMenuShortcut>False</StartMenuShortcut>
</WAUConfig>
"@

    $xmlContent | Out-File -FilePath $ConfigPath -Encoding UTF8 -Force
    Write-Log "WAUConfig.xml created successfully"

    Write-Log "Deployment completed successfully"

} catch {
    Write-Log "ERROR: $($_.Exception.Message)"
    exit 1
} finally {
    # Cleanup
    if (Test-Path $TempPath) {
        Remove-Item -Path $TempPath -Recurse -Force -ErrorAction SilentlyContinue
    }
}

**Reply by KnifMelti** [reply]
We don't use any WAUConfig.xml...
You can use properties when installing the MSI, look in:
https://github.com/Romanitho/Winget-AutoUpdate?tab=readme-ov-file#advanced-installation

**Reply by Champ895** [reply]
#Requires -RunAsAdministrator

param(
    [string[]]$BlacklistedApps = @(
        "Microsoft.Office.Desktop",
        "NitroSoftware.NitroPro"
    )
)

$TempPath = "$env:TEMP\WAU_Deploy"
$GitHubRepo = "Romanitho/Winget-AutoUpdate"
$LogFile = "$env:ProgramData\WAU_Deploy.log"

function Write-Log {
    param([string]$Message)
    $Timestamp = Get-Date -Format "yyyy-MM-dd HH:mm:ss"
    "$Timestamp - $Message" | Tee-Object -FilePath $LogFile -Append
}

try {
    Write-Log "Starting Winget-AutoUpdate deployment"

    # Create temp directory
    if (!(Test-Path $TempPath)) {
        New-Item -ItemType Directory -Path $TempPath -Force | Out-Null
    }

    # Get latest release info from GitHub API
    Write-Log "Fetching latest release information from GitHub"
    $ApiUrl = "https://api.github.com/repos/$GitHubRepo/releases/latest"
    $Release = Invoke-RestMethod -Uri $ApiUrl -UseBasicParsing

    # Find MSI download URL
    $MsiAsset = $Release.assets | Where-Object { $_.name -like "*.msi" }
    if (!$MsiAsset) {
        throw "MSI file not found in latest release"
    }

    $MsiUrl = $MsiAsset.browser_download_url
    $MsiPath = Join-Path $TempPath $MsiAsset.name

    Write-Log "Downloading MSI: $($MsiAsset.name)"
    Invoke-WebRequest -Uri $MsiUrl -OutFile $MsiPath -UseBasicParsing

    # Prepare MSI property string for install
    # Based on: https://github.com/Romanitho/Winget-AutoUpdate#advanced-installation
    $BlacklistString = ($BlacklistedApps -join ",")
    $msiProps = @(
        "BLACKLIST=$BlacklistString",
        "UPDATEFREQUENCY=Daily",
        "UPDATETIME=12:00",
        "UPDATEATLOGON=1",
        "REINSTALLONPOLICYUPDATE=1",
        "NOTIFICATIONLEVEL=SuccessOnly",
        "DESKTOPSHORTCUT=0",
        "STARTMENUSHORTCUT=0"
    ) -join " "

    Write-Log "Installing Winget-AutoUpdate with properties: $msiProps"
    $InstallArgs = "/i `"$MsiPath`" /quiet /norestart $msiProps"
    $Process = Start-Process -FilePath "msiexec.exe" -ArgumentList $InstallArgs -Wait -PassThru

    if ($Process.ExitCode -ne 0) {
        throw "Installation failed with exit code: $($Process.ExitCode)"
    }

    Write-Log "Deployment completed successfully"

} catch {
    Write-Log "ERROR: $($_.Exception.Message)"
    exit 1
} finally {
    # Cleanup
    if (Test-Path $TempPath) {
        Remove-Item -Path $TempPath -Recurse -Force -ErrorAction SilentlyContinue
    }
}

**Reply by KnifMelti** [reply]
Where you found REINSTALLONPOLICYUPDATE=1 I don't know...
...just try it out yourself.

**Reply by Champ895** [reply]
Looks like it's working so far so good with a manual run.

**Reply by Champ895** [reply]
msi exclusion does not work. I guess I need to include the text file instead for excluded apps.
